Allow a service account installation

smarterclayton opened the pull request:
Allow a service account to be used as the installation target. Can
potentially simplify metrics and logging installation.

@liggitt review please, low impact to other components but potential
better story for integrators

smarterclayton commented:
[test]

smarterclayton commented:
Any feedback on this? @csrwng as well

csrwng commented on pkg/cmd/cli/cmd/newapp.go:
```diff
@@ -233,13 +233,14 @@ func RunNewApplication(fullName string, f *clientcmd.Factory, out io.Writer, c *
 		afterFn = configcmd.NewPrintNameOrErrorAfterIndent(mapper, shortOutput, "created", out, c.Out(), indent)
 	default:
 		afterFn = configcmd.NewPrintErrorAfter(mapper, c.Out())
+		afterFn = configcmd.HaltOnError(afterFn)
```
do we really want to halt on error for normal app creation? Until now if you already have an existing resource we say we couldn't create it but still create all the other ones. I guess it's ok if we change the behavior, just want to make sure we're making a conscious decision to do so.

csrwng replied:
nm, was reading this backwards. please ignore :)

smarterclayton replied:
In this case all of the security stuff is first, so if we get an error
creating service account we probably don't want to create the pod.  Right
now we only allow a single installable component.

On Nov 5, 2015, at 4:59 PM, Cesar Wong <notifications@github.com> wrote:

In pkg/cmd/cli/cmd/newapp.go
<https://github.com/openshift/origin/pull/5724#discussion_r44075605>:

>  	switch {
>  	// only print success if we don't have installables
>  	case !result.GeneratedJobs:
>  		afterFn = configcmd.NewPrintNameOrErrorAfterIndent(mapper, shortOutput, "created", out, c.Out(), indent)
>  	default:
>  		afterFn = configcmd.NewPrintErrorAfter(mapper, c.Out())
> +		afterFn = configcmd.HaltOnError(afterFn)

do we really want to halt on error for normal app creation? Until now if
you already have an existing resource we say we couldn't create it but
still create all the other ones. I guess it's ok if we change the behavior,
just want to make sure we're making a conscious decision to do so.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/5724/files#r44075605>.

csrwng commented on pkg/cmd/cli/cmd/newapp.go:
```diff
@@ -531,6 +533,12 @@ func handleRunError(c *cobra.Command, err error, fullName string) error {
 	switch t := err.(type) {
 	case newcmd.ErrRequiresExplicitAccess:
+		if t.Input.Token != nil && t.Input.Token.ServiceAccount {
+			return fmt.Errorf(`installing %q requires that you create the 'installer' service account with project admin access
```
s/admin/edit/

csrwng commented:
other than one comment above, lgtm

smarterclayton commented:
Added support for setting env on the installable (parameterization) and also always generating POD_NAMESPACE for now as a workaround so oc in the container knows where it is.  Added a few more tests, but post 3.1 I will add even more tests.

smarterclayton commented:
@deads2k view the second commit, @csrwng this has all of your suggested changes and better messages

csrwng commented:
@smarterclayton the service account is still not getting the namespace set... 
```
12:54 # oc get rolebindings
NAME                     ROLE                    USERS                              GROUPS                               SERVICE ACCOUNTS   SUBJECTS
system:image-pullers     /system:image-puller                                       system:serviceaccounts:testinstall
admins                   /admin                  cewong
installer-role-binding   /edit                   system:serviceaccount::installer
system:deployers         /system:deployer                                                                                deployer
system:image-builders    /system:image-builder                                                                           builder
```

I also see this error when SA is getting created along with the pod:

```
error: Pod "myinstaller-install" is forbidden: no API token found for service account testinstall/installer, retry after the token is automatically created and added to the service account
```

deads2k commented:
In person explanation convinced me that invalid input getting weird results in an ok thing.  I think that validation ensuring that non-nil fields matched would be desirable, but the second commit lgtm.

csrwng commented:
Sorry @smarterclayton. Please ignore my previous comment. First I was running without having restarted the server with the new code, then I was using a project where the SA already existed. I blew away everything and now it works.

csrwng commented on pkg/cmd/cli/cmd/newapp.go:
```diff
@@ -531,8 +533,20 @@ func handleRunError(c *cobra.Command, err error, fullName string) error {
+
+You can see more information about the image by adding the --dry-run flag.
+If you trust the provided image, include the flag --grant-install-rights.`, t.Match.Value)
+		}
```
@smarterclayton just one more nit about this message. It'd be nice if it told me that if I specify --grant-install-rights, a service account will be created for me if one doesn't already exist.

smarterclayton replied:
Let's queue that up for after - I agree though.  Is that ok?

On Fri, Nov 6, 2015 at 2:02 PM, Cesar Wong <notifications@github.com> wrote:

> In pkg/cmd/cli/cmd/newapp.go
> <https://github.com/openshift/origin/pull/5724#discussion_r44174382>:
>
> > @@ -531,8 +533,20 @@ func handleRunError(c *cobra.Command, err error, fullName string) error {
> >  	}
> >  	switch t := err.(type) {
> >  	case newcmd.ErrRequiresExplicitAccess:
> > +		if t.Input.Token != nil && t.Input.Token.ServiceAccount {
> > +			return fmt.Errorf(`installing %q requires an 'installer' service account with project editor access
> > +
> > +WARNING: This will allow the pod to create and manage resources within your namespace -
> > +ensure you trust the image with those permissions before you continue.
> > +
> > +You can see more information about the image by adding the --dry-run flag.
> > +If you trust the provided image, include the flag --grant-install-rights.`, t.Match.Value)
> > +		}
>
> @smarterclayton <https://github.com/smarterclayton> just one more nit
> about this message. It'd be nice if it told me that if I specify
> --grant-install-rights, a service account will be created for me if one
> doesn't already exist.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/5724/files#r44174382>.
>

csrwng replied:
yup

smarterclayton commented:
[merge]

csrwng commented:
@smarterclayton I still occasionally see this:
```
error: Pod "myinstaller-install" is forbidden: no API token found for service account devproj/installer, retry after the token is automatically created and added to the service account
```

It looks like the create should happen in 2 stages.

smarterclayton commented:
Guess we have to retry in a few cases - sucks

On Nov 6, 2015, at 2:50 PM, Cesar Wong <notifications@github.com> wrote:

@smarterclayton <https://github.com/smarterclayton> I still occasionally
see this:

error: Pod "myinstaller-install" is forbidden: no API token found for
service account devproj/installer, retry after the token is
automatically created and added to the service account

It looks like the create should happen in 2 stages.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/5724#issuecomment-154514463>.
